### ISSUE TITLE
Use of deprecated OSAtomic* functions

### ISSUE THREAD

**Issue by equinox1993** [sctplab/usrsctp#177]
Hi,

We are using the library in iOS and have recently decided to bump up our deployment target to iOS 10.0. However, we started getting errors for use of deprecated functions:

    ../../third_party/usrsctp/usrsctplib/usrsctplib/netinet/sctp_usrreq.c:861:7: error: 'OSAtomicCompareAndSwapIntBarrier' is deprecated: first deprecated in iOS 10.0 - Use atomic_compare_exchange_strong() from <stdatomic.h> instead [-Werror,-Wdeprecated-declarations]
                (atomic_cmpset_int(&inp->sctp_flags, flags, (flags | SCTP_PCB_FLAGS_SOCKET_GONE | SCTP_PCB_FLAGS_CLOSE_IP)))) {
                 ^
    ../../third_party/usrsctp/usrsctplib/usrsctplib/user_atomic.h:59:42: note: expanded from macro 'atomic_cmpset_int'
    #define atomic_cmpset_int(dst, exp, src) OSAtomicCompareAndSwapIntBarrier(exp, src, (int *)dst)
                                             ^
    /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator11.0.sdk/usr/include/libkern/OSAtomicDeprecated.h:580:6: note: 'OSAtomicCompareAndSwapIntBarrier' has been explicitly marked deprecated here
    bool    OSAtomicCompareAndSwapIntBarrier( int __oldValue, int __newValue, volatile int *__theValue );
            ^
    ../../third_party/usrsctp/usrsctplib/usrsctplib/netinet/sctp_usrreq.c:1150:6: error: 'OSAtomicAdd32Barrier' is deprecated: first deprecated in iOS 10.0 - Use atomic_fetch_add() from <stdatomic.h> instead [-Werror,-Wdeprecated-declarations]
                                            SCTP_STAT_INCR_COUNTER32(sctps_aborted);
                                            ^
    ../../third_party/usrsctp/usrsctplib/usrsctplib/netinet/sctp_uio.h:1130:38: note: expanded from macro 'SCTP_STAT_INCR_COUNTER32'
    #define SCTP_STAT_INCR_COUNTER32(_x) SCTP_STAT_INCR(_x)
                                         ^
    ../../third_party/usrsctp/usrsctplib/usrsctplib/netinet/sctp_uio.h:1120:28: note: expanded from macro 'SCTP_STAT_INCR'
    #define SCTP_STAT_INCR(_x) SCTP_STAT_INCR_BY(_x,1)
                               ^
    ../../third_party/usrsctp/usrsctplib/usrsctplib/netinet/sctp_uio.h:1126:34: note: expanded from macro 'SCTP_STAT_INCR_BY'
    #define SCTP_STAT_INCR_BY(_x,_d) atomic_add_int(&SCTP_BASE_STAT(_x), _d)
                                     ^
    ../../third_party/usrsctp/usrsctplib/usrsctplib/user_atomic.h:56:35: note: expanded from macro 'atomic_add_int'
    #define atomic_add_int(addr, val) OSAtomicAdd32Barrier(val, (int32_t *)addr)
                                      ^
    /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator11.0.sdk/usr/include/libkern/OSAtomicDeprecated.h:161:9: note: 'OSAtomicAdd32Barrier' has been explicitly marked deprecated here
    int32_t OSAtomicAdd32Barrier( int32_t __theAmount, volatile int32_t *__theValue );
            ^
    ../../third_party/usrsctp/usrsctplib/usrsctplib/netinet/sctp_usrreq.c:1155:6: error: 'OSAtomicAdd32Barrier' is deprecated: first deprecated in iOS 10.0 - Use atomic_fetch_add() from <stdatomic.h> instead [-Werror,-Wdeprecated-declarations]
                                            SCTP_STAT_DECR_GAUGE32(sctps_currestab);
                                            ^
    ../../third_party/usrsctp/usrsctplib/usrsctplib/netinet/sctp_uio.h:1135:36: note: expanded from macro 'SCTP_STAT_DECR_GAUGE32'
    #define SCTP_STAT_DECR_GAUGE32(_x) SCTP_STAT_DECR(_x)
                                       ^
    ../../third_party/usrsctp/usrsctplib/usrsctplib/netinet/sctp_uio.h:1121:28: note: expanded from macro 'SCTP_STAT_DECR'
    #define SCTP_STAT_DECR(_x) SCTP_STAT_DECR_BY(_x,1)
                               ^
    ../../third_party/usrsctp/usrsctplib/usrsctplib/netinet/sctp_uio.h:1127:34: note: expanded from macro 'SCTP_STAT_DECR_BY'
    #define SCTP_STAT_DECR_BY(_x,_d) atomic_subtract_int(&SCTP_BASE_STAT(_x), _d)
                                     ^
    ../../third_party/usrsctp/usrsctplib/usrsctplib/user_atomic.h:58:40: note: expanded from macro 'atomic_subtract_int'
    #define atomic_subtract_int(addr, val) OSAtomicAdd32Barrier(-val, (int32_t *)addr)
                                           ^
    /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator11.0.sdk/usr/include/libkern/OSAtomicDeprecated.h:161:9: note: 'OSAtomicAdd32Barrier' has been explicitly marked deprecated here
    int32_t OSAtomicAdd32Barrier( int32_t __theAmount, volatile int32_t *__theValue );
            ^

Basically OSAtomicAdd32Barrier and OSAtomicCompareAndSwapIntBarrier are deprecated on iOS 10, which are used in user_atomic.h

We are using our own [build rule](https://cs.chromium.org/chromium/src/third_party/usrsctp/BUILD.gn) and we are a bit hesitate to just suppress the warning instead of fixing the bug. I wonder if you guys have a long term plan to replace them?

The suggestion from the compiler is to use atomic_fetch_add() from <stdatomic.h>, but that would require some nontrivial refactoring in the codebase. We will need to use std::atomic instead of raw pointer.

**Comment by tuexen** [comment]
The long term plan is to use the functions you suggest on newer platforms. Right now we suppress the warnings in the build scripts provided. But it will be fixed eventually.

**Comment by equinox1993** [comment]
Cool. So for now I think we could simply suppress that warning in our build script. Do you have an estimated time of moving to stdatomic?

**Comment by tuexen** [comment]
Not really. Fixing some bugs has definitely higher priority...

**Comment by tkoeppe** [comment]
There's perhaps some additional opportunity cost here: the lack of standard atomics means that dynamic debuggers such as TSAN cannot diagnose through any code that uses usrsctp, and thus we're missing out on valuable debugging of concurrency code.

After a brief look, I wonder how hard it would be to add support for standard atomics: you'd need to update (or add to) the macros in user_atomic.h, and then make all the variables that are accessed via those macros conditionally `_Atomic` (e.g. create a new macro `USRSCTP_ATOMIC` to wrap those variable/member definitions). Most of the uses of the atomic macros seems to go through refcounting logic, so maybe this wouldn't actually be such a large change?

**Comment by tuexen** [comment]
Please note that this code is also used also in kernel land. So the macros must abstract that too.

**Comment by tkoeppe** [comment]
For what it's worth, I've prototypes this change; it only took a few minutes. If you're interested I can post a patch.

**Comment by tuexen** [comment]
Sure. We can look at it.

**Comment by tkoeppe** [comment]
Here you go:

[usrsctp-atomics.patch.txt](https://github.com/sctplab/usrsctp/files/3569525/usrsctp-atomics.patch.txt)

All the `_Atomic`s would have to be turned into macro invocations.

The `atomic_init` function should probably just be renamed regardless (it clashes with a C standard library function).

For compare-and-exchange, the expected operand needs to be an address, so we need to have variables for those values.

**Comment by tuexen** [comment]
How does this work on platforms not support C11? How does this work in kernel land? How to deal with changing `struct sctpstat`, which is part of the API?

**Comment by tkoeppe** [comment]
I would add this code as a new, conditionally-compiled mode, in addition to the one that uses GCC intrinsics, so that this feature would only be used on demand (e.g. in a project that has C11 available and only runs in userspace). Existing code can continue as is. Is that reasonable?

I think if you really pressed the issue, you could squint at it decide that `_Atomic int` and `int` are layout compatible on the platforms on which you want to interop Linux kernel and userspace code (i.e. on Linux, with GCC), and that the relevant synchronization primitives are interoperable, too. But I'd also be happy if people who talk to the kernel would _not_ use this new mode.

**Comment by tuexen** [comment]
> I would add this code as a new, conditionally-compiled mode, in addition to the one that uses GCC intrinsics, so that this feature would only be used on demand (e.g. in a project that has C11 available and only runs in userspace). Existing code can continue as is. Is that reasonable?

That sounds reasonable. You can create a PR, if you want...

> 
> I think if you really pressed the issue, you could squint at it decide that `_Atomic int` and `int` are layout compatible on the platforms on which you want to interop Linux kernel and userspace code (i.e. on Linux, with GCC), and that the relevant synchronization primitives are interoperable, too. But I'd also be happy if people who talk to the kernel would _not_ use this new mode.
> 

The sources in `netinet/` and `netinet6` also live in the FreeBSD kernel (and could be compiled as a NKE for Mac OS). So not changing that code would be preferable. One could use `#ifdef` here.

> There would already be value in this feature for people who write pure userspace code and want to sanitize it for concurrency errors.

**Comment by tkoeppe** [comment]
Hi, is the constraint on `netinet/` still valid? I have draft changes in https://github.com/tkoeppe/usrsctp/commit/169c9ec7c514b5e2018ae64f69ceccdab1e18f31, but if we need to `#ifdef` everything, it'll be a bit more work.

**Comment by bdaehlie** [comment]
I'd still love to see this get fixed. There is a patch from 2019 earlier in this thread, but the patch doesn't seem to target what I'd expect. Maybe it's a matter of just using atomic_fetch_add and atomic_compare_exchange_strong in these macros, instead of OSAtomic* stuff?

#define atomic_add_int(addr, val) OSAtomicAdd32Barrier(val, (int32_t *)addr)
#define atomic_fetchadd_int(addr, val) OSAtomicAdd32Barrier(val, (int32_t *)addr)
#define atomic_subtract_int(addr, val) OSAtomicAdd32Barrier(-val, (int32_t *)addr)
#define atomic_cmpset_int(dst, exp, src) OSAtomicCompareAndSwapIntBarrier(exp, src, (int *)dst)

Unfortunately I don't understand this stuff enough to fix it myself with any confidence.

The deprecation warnings from these are a huge percentage of the build warnings for Firefox on macOS.

**Comment by bdaehlie** [comment]
My last comment was incorrect, someone explained to me what I was missing. Looks like Firefox is not going to be using this for much longer so my interest in fixing this is diminished as a result, but it would still be a nice fix for other users.
